h/models/group.py

```diff
@@ -221,56 +222,67 @@ def is_public(self):
 
-        write_principal = _write_principal(self)
-        if write_principal is not None:
-            terms.append((security.Allow, write_principal, Permission.Group.WRITE))
+        # Any logged in user should be able to flag things they can see
+        if self.readable_by == ReadableBy.members:
+            terms.append((Allow, in_group_principal, Permission.Group.FLAG))
+        elif self.readable_by == ReadableBy.world:
+            terms.append((Allow, security.Authenticated, Permission.Group.FLAG))
+
+        if self.writeable_by == WriteableBy.authority:
+            terms.append((Allow, in_authority_principal, Permission.Group.WRITE))
+        elif self.writeable_by == WriteableBy.members:
+            terms.append((Allow, in_group_principal, Permission.Group.WRITE))
 
         if self.creator:
-            # The creator of the group should be able to update it
-            terms.append((security.Allow, self.creator.userid, Permission.Group.ADMIN))
-            terms.append(
-                (security.Allow, self.creator.userid, Permission.Group.MODERATE)
-            )
+            terms.append((Allow, self.creator.userid, Permission.Group.MODERATE))
             # The creator may update this group in an upsert context
-            terms.append((security.Allow, self.creator.userid, Permission.Group.UPSERT))
-
-        # This authority principal may be used to grant auth clients
-        # permissions for groups within their authority
-        authority_principal = "client_authority:{}".format(self.authority)
+            terms.append((Allow, self.creator.userid, Permission.Group.UPSERT))
 
-        # auth_clients that have the same authority as the target group
-        # may read the members within it
-        terms.append(
-            (security.Allow, authority_principal, Permission.Group.MEMBER_READ)
-        )
         # auth_clients that have the same authority as the target group
         # may add members to it
-        terms.append((security.Allow, authority_principal, Permission.Group.MEMBER_ADD))
+        terms.append((Allow, client_authority_principal, Permission.Group.MEMBER_ADD))
+
+        # Read permissions ------------------------------------------------ #
+
+        if self.readable_by == ReadableBy.members:
+            terms.append((Allow, in_group_principal, Permission.Group.READ))
+            terms.append((Allow, in_group_principal, Permission.Group.MEMBER_READ))
+        elif self.readable_by == ReadableBy.world:
+            terms.append((Allow, security.Everyone, Permission.Group.READ))
+            terms.append((Allow, security.Everyone, Permission.Group.MEMBER_READ))
+
+        # auth_clients with matching authority should be able to read the group
+        # and it's members
+        terms.append((Allow, client_authority_principal, Permission.Group.READ))
+        terms.append((Allow, client_authority_principal, Permission.Group.MEMBER_READ))
```

Comment:
These have also been re-arranged so you can see all the circumstances in which the same permissions are granted. In this case, honestly, the `MEMBER_READ` permission is just the `READ` permission, and we might want to consider making them the same thing.